lib/collection/src/shards/shard_holder.rs

```diff
@@ -67,28 +70,65 @@ pub enum ShardHashRing {
 }
 
 impl ShardHashRing {
-    /// Create a new single hashring
+    /// Create a new single hashring.
     ///
     /// The hashring is created with a fair distribution of points and `HASH_RING_SHARD_SCALE` scale.
     pub fn single() -> Self {
         Self::Single(HashRing::fair(HASH_RING_SHARD_SCALE))
     }
 
+    /// Create a new resharding hashring, with resharding shard already added into `new` hashring.
+    ///
+    /// The hashring is created with a fair distribution of points and `HASH_RING_SHARD_SCALE` scale.
+    pub fn resharding(shard_id: ShardId) -> Self {
+        let mut ring = Self::Resharding {
+            old: HashRing::fair(HASH_RING_SHARD_SCALE),
+            new: HashRing::fair(HASH_RING_SHARD_SCALE),
+        };
+
+        ring.add_resharding(shard_id);
+
+        ring
+    }
+
     pub fn is_empty(&self) -> bool {
         match self {
             Self::Single(ring) => ring.is_empty(),
             Self::Resharding { old, new } => old.is_empty() && new.is_empty(),
         }
     }
 
+    pub fn is_resharding(&self) -> bool {
+        matches!(self, Self::Resharding { .. })
+    }
+
     pub fn add(&mut self, shard: ShardId) {
         match self {
             Self::Single(ring) => ring.add(shard),
-            // When resharding, update the new ring
-            Self::Resharding { old: _, new } => new.add(shard),
+            Self::Resharding { old, new } => {
+                if new.get(&shard).is_none() {
+                    old.add(shard);
+                    new.add(shard);
+                }
+            }
```

Comment:
So, this design is the simplest way to make `ShardHashRing::add` work properly with `ShardHolder::load_shards`/`ShardHolder::add_shard` with minimal changes.

With this implementation, we can add our resharding shards to the hashring (with `ShardHashRing::add_resharding`) *before* calling `load_shards`/`add_shard`.

And then, we can just call `load_shards`/`add_shard` as usual (without any changes to signature/logic/order of calls/etc) and it will properly add all shards to *both* hashrings, and simply "skip" any hashring changes for the resharding shards that are already in the hashring.